documentation/design-docs/ipc-protocol.md

```diff
@@ -0,0 +1,555 @@
+
+### General Flow
+
+All communication with the Diagnostic Server will begin with a Diagnostic IPC Message sent from the client to the server.  The server will respond with a Diagnostic IPC Message.  After this, the client and runtime _may_ reuse the Pipe for any Command specific communication which is referred to as an Optional Continuation.
+
+![Generic Flow](ipc-protocol-genericflow.svg)
+
+```
+runtime <- client : [ Header ][ Payload ]
+runtime -> client : [ Header ][ Payload ]
+optional:
+    runtime <-> client : [ Optional Continuation ]
+connection closed
+```
+
+Example flow for EventPipe:
+```
+runtime <- client : [ ver; size; EventPipe CollectTracing ][ stream config struct  ] <- Diagnostic IPC Message
+runtime -> client : [ ver; size; EventPipe TracingStarted ][ stream started struct ] <- Diagnostic IPC Message
+runtime -> client : [ stream of netperf data ]                                       <- Optional Continuation
+
+// stop message is sent on another connection
+
+connection closed
+```
+
+Example flow for hypothetical application with custom communication protocol:
+```
+runtime <- client : [ ver; size; start_app_session_command    ][ app config struct      ] <- Diagnostic IPC Message
+runtime -> client : [ ver; size; app_session_started_response ][ session started struct ] <- Diagnostic IPC Message
+
+Optional Continuation using app specific protocol
+
+runtime <- client : [ command specific protocol ]                                         <- Optional Continuation
+runtime -> client : [ command specific protocol ]                                         <- Optional Continuation
+Eventually...
+runtime <- client : [ command specific protocol (close connection) ]                      <- Optional Continuation
+connection closed
+```
+
+### Transport
+
+The protocol will be communicated over a platform-specific transport.  On Unix/Linux based platforms, a Unix Domain Socket will be used, and on Windows, a Named Pipe will be used.
+
+#### Naming and Location Conventions
+
+Unix Domain Sockets (MacOS and *nix):
+
+The location of the socket is determined by a call to `GetTempPathA()` in the PAL.  This has a fallthrough behavior and attempts to use the following three locations, in order:
+1. (Mac only) App Group Container if `AppGroupId` is not null
+2. The directory stored in the `$TMPDIR` environment variable
+3. `/tmp`
+
+In order to ensure filename uniqueness, a `disambiguation key` is generated.  On Mac and NetBSD, this is the process start time encoded as the number of seconds since UNIX epoch time.  If `/proc/$PID/stat` is available (all other *nix platforms), then the process start time encoded as jiffies since boot time is used.
+
+socket name:
+```c
+dotnetcore-diagnostic-{%d:PID}-{%llu:disambiguation key}-socket
```

Comment:
Are we going to need a sub-directory to get more control on the file permissions or you determined that we definately won't need one? If we are uncertain we could err on the side of caution and set the path as \<TEMP>/dotnet-{%d:PID}-{%llu:disambiguation key}/diagnostic-socket. If it winds up we didn't need the directory it doesn't seem like much overhead.

Reply:
Separately I recommend 'dotnetcore' -> 'dotnet' given our branding message that "Core" is going away and to align with "DOTNET" used in the magic string.

Reply:
I'll do some more verification to ensure we are getting the perms we want on the file.  If we decide to go with having a directory, I think
```
/tmp/com.microsoft.dotnet/dotnet-diagnostic-{%d}-{%llu}.sock
```
would give us a more reusable system.  Other pipes, locks, etc. could live that parent directory as well.  (this name scheme is similar to what Office does on Mac)